### PR TITLE
LITE-27889 Add Configuration and File models

### PR DESCRIPTION
**Pull request by jonatrios**

### CODE REVIEW

File: connect_ext_ppr/models/enums.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import enum
+
+
+class DeploymentStatusChoices(str, enum.Enum):
+    PENDING = 'pending'
+    PROCESSING = 'processing'
+    SYNCED = 'synced'
+
+
+class MimeTypeChoices(str, enum.Enum):
```
Comment (rahulmondal) [inline connect_ext_ppr/models/enums.py:10]: This enum is not related to models. Any specific reason for putting this under model enums?

File: connect_ext_ppr/events.py [hunk 2]
```diff
@@ -42,7 +43,7 @@ def handle_listing_processing(self, request):
         if request['status'] == 'listed':
             mp = get_marketplaces(self.client, [request['contract']['marketplace']['id']]).first()
```
Comment (rahulmondal) [inline connect_ext_ppr/events.py:44]:
```suggestion
            mp = get_marketplaces(self.installation_client, [request['contract']['marketplace']['id']]).first()
```

File: connect_ext_ppr/models/utils.py [hunk 3]
```diff
@@ -4,35 +4,35 @@
 
 def add_deployments(installation, listings, config, logger):
```
Comment (rahulmondal) [inline connect_ext_ppr/models/utils.py:5]: I think we should move this to a service file. It's more like a service than a utility.

File: connect_ext_ppr/models/utils.py [hunk 4]
```diff
@@ -4,35 +4,35 @@
...
     with get_db_ctx_manager(config) as db:
-        print(db.bind)
         deployments = []
+        seen = set()
         for li in listings:
             product_id = li['product']['id']
             for hub in li['contract']['marketplace']['hubs']:
```
Comment (rahulmondal) [inline connect_ext_ppr/models/utils.py:11]: I am not sure that the listing object provides hub details under the marketplace property.
image

File: connect_ext_ppr/events.py [hunk 5]
```diff
@@ -42,7 +43,7 @@ def handle_listing_processing(self, request):
...
             request['contract']['marketplace'] = mp
```
Comment (rahulmondal) [inline connect_ext_ppr/events.py:45]: There is a possibility to reuse the add_deployments function here. Here we are doing it for a single listing and in add_deployments we are doing the same action for a list of listings.

File: connect_ext_ppr/events.py [hunk 6]
```diff
@@ -42,7 +43,7 @@ def handle_listing_processing(self, request):
         if request['status'] == 'listed':
```
Comment (rahulmondal) [inline connect_ext_ppr/events.py:43]: As we are not doing anything with 'unlisted' event, it's better not to subscribe to it at all.

Same with 'uninstalled' event of 'installation_status_change'

**Review by rahulmondal — CHANGES_REQUESTED** [review]

File: connect_ext_ppr/models/enums.py (hunk 1 above)
Comment (jonatrios) [reply connect_ext_ppr/models/enums.py:10]: Yes, as the mime_type of the files we are going allow the user to upload are basically json and excel, for me was a good idea to limit options.

File: connect_ext_ppr/models/utils.py (hunk 4 above)
Comment (jonatrios) [reply connect_ext_ppr/models/utils.py:11]: Not but at this point I have all information needed processed by `get_all_info` function, I inject hub information there

File: connect_ext_ppr/models/configuration.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    )
+    created_at = db.Column(db.DateTime(), default=datetime.utcnow)
+    created_by = db.Column(db.String(20))
+    updated_at = db.Column(db.DateTime(), default=datetime.utcnow)
+    updated_by = db.Column(db.String(20))
+
+    def activate(self):
+        self.state = ConfigurationStateChices.ACTIVE
+
+    def deleted(self):
```
Comment (Hairash) [inline connect_ext_ppr/models/configuration.py:31]: [COD-C] Maybe better `delete` as it is an action, WDYT?

File: connect_ext_ppr/models/enums.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    application_json = 'application/json'
```
Comment (Hairash) [inline connect_ext_ppr/models/enums.py:11]: [COD-C] Just an idea - to write names of variants capitalized for unification, might it be convenient?

File: connect_ext_ppr/models/utils.py [hunk 9]
```diff
@@ -4,35 +4,35 @@
...
+                comb = (product_id, installation['owner']['id'], hub['hub']['id'])
                 q = db.query(Deployment).filter_by(
```
Comment (Hairash) [inline connect_ext_ppr/models/utils.py:13]: [COD-C] I propose to check `if comb not in seen:` there to skip the query below.
Also I propose to write the following way:
```
if comb in seen:
    continue
dep = Deployment(
...
```
to reduce indentation level.
Do you agree with that?

File: connect_ext_ppr/events.py (hunk 2 above)
Comment (jonatrios) [reply connect_ext_ppr/events.py:44]: Good catch!

File: connect_ext_ppr/events.py (hunk 5 above)
Comment (jonatrios) [reply connect_ext_ppr/events.py:45]: Agree, lets do it!

File: connect_ext_ppr/models/utils.py [hunk 10]
```diff
@@ -4,35 +4,35 @@
...
-                deployments.append(dep)
+                if comb not in seen:
+                    dep = Deployment(
+                        product_id=product_id,
+                        hub_id=hub['hub']['id'],
+                        vendor_id=li['vendor']['id'],
+                        account_id=installation['owner']['id'],
+                    )
+                    logger.info(
+                        f"Generating Deployment: product={product_id},"
```
Comment (Hairash) [inline connect_ext_ppr/models/utils.py:28]: [COD-C] Let's use single quotes?

**Review by Hairash — APPROVED** [review]

File: connect_ext_ppr/models/utils.py (hunk 10 above)
Comment (jonatrios) [reply connect_ext_ppr/models/utils.py:28]: One convention that we have about this on the connect is to use double quotes for strings displayed for the user, I think is not an issue to maintain this for extensions.

Comment (Hairash) [reply connect_ext_ppr/models/utils.py:28]: Ok

File: connect_ext_ppr/models/configuration.py (hunk 7 above)
Comment (jonatrios) [reply connect_ext_ppr/models/configuration.py:31]: If you refer to the article it said that for deletion, should be a mechanism of fake delete, that will mark configuration file as deleted, because it could be related to an "in use" PPR build on top of it, so my method is more like "mark as deleted", not delete like the ORM will do.

File: connect_ext_ppr/models/enums.py (hunk 8 above)
Comment (jonatrios) [reply connect_ext_ppr/models/enums.py:11]: For mime types, media service (that is where I take this), is doing it this way, maybe it not a problem for this Emun to keep this way, and for status capital keys, wdyt?

File: connect_ext_ppr/models/utils.py (hunk 9 above)
Comment (jonatrios) [reply connect_ext_ppr/models/utils.py:13]: But `if comb not in seen` does not know if the object already exists in the DB, here we could have:

1. Combination exists in DB: for loop will continue
2. Combination does not exists in DB: `set` will be in charge of define if the combination that is going to be part of bulk create list already exists on the set of non existing DB objects.

**Review by rahulmondal — APPROVED** [review]
